### ISSUE TITLE
Loss remains constant during training

### ISSUE THREAD

**Issue by wilfredkisku** [Xilinx/brevitas#399]
I am using brevitas for the very first time and I need help. I am trying to train a quantized model, where just the filter weights are quantized to a fixed point 4/8/16 bit representation. The below model is the same that I have created for training.
```
import brevitas.nn as qnn

# define the CNN architecture
class QNet(nn.Module):
    def __init__(self):
        super(QNet, self).__init__()
        # convolutional layer (sees 32x32x3 image tensor)
        self.conv1 = qnn.QuantConv2d(3, 16, 3, weight_bit_width=8, padding=1)
        # convolutional layer (sees 16x16x16 tensor)
        self.conv2 = qnn.QuantConv2d(16, 32, 3, weight_bit_width=8, padding=1)
        # convolutional layer (sees 8x8x32 tensor)
        self.conv3 = qnn.QuantConv2d(32, 64, 3, weight_bit_width=8, padding=1)
        # max pooling layer
        self.pool = nn.MaxPool2d(2, 2)
        # linear layer (64 * 4 * 4 -> 500)
        self.fc1 = qnn.QuantLinear(64 * 4 * 4, 500, bias = True, weight_bit_width=8)
        # linear layer (500 -> 10)
        self.fc2 = qnn.QuantLinear(500, 10, bias = True, weight_bit_width=8)
        # dropout layer (p=0.25)
        #self.dropout = nn.Dropout(0.25)

    def forward(self, x):
        # add sequence of convolutional and max pooling layers
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = self.pool(F.relu(self.conv3(x)))
        # flatten image input
        x = x.view(-1, 64 * 4 * 4)
        # add dropout layer
        #x = self.dropout(x)
        # add 1st hidden layer, with relu activation function
        x = F.relu(self.fc1(x))
        # add dropout layer
        #x = self.dropout(x)
        # add 2nd hidden layer, with relu activation function
        x = self.fc2(x)
        return x

# create a complete CNN
qnet = QNet()
print(qnet)

# move tensors to GPU if CUDA is available
if train_on_gpu:
    qnet.cuda()
```
During training I see that the loss does not change. Can anyone point out my mistake please?
```
Epoch: 1 	Training Loss: 2.302958 	Validation Loss: 2.302216
Validation loss decreased (inf --> 2.302216).  Saving model ...
Epoch: 2 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 3 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 4 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 5 	Training Loss: 2.302958 	Validation Loss: 2.302216
Validation loss decreased (2.302216 --> 2.302216).  Saving model ...
Epoch: 6 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 7 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 8 	Training Loss: 2.302958 	Validation Loss: 2.302216
Validation loss decreased (2.302216 --> 2.302216).  Saving model ...
Epoch: 9 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 10 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 11 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 12 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 13 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 14 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 15 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 16 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 17 	Training Loss: 2.302958 	Validation Loss: 2.302216
Epoch: 18 	Training Loss: 2.302958 	Validation Loss: 2.302216
```

Thank you.

**Comment by MohamedA95** [comment]
HI, 
Try to disable all sorts of quantization first to pinpoint the source of the issue, is it the structure or the quantization?

**Comment by wilfredkisku** [comment]
Thanks for the reply. I had trained the model normally (without any quantization wrapper) before I added them and it had worked fine, with a validation accuracy of around 77%.

**Comment by MohamedA95** [comment]
Just to make sure we are on the same page you trained the model using brevitas layers i.e `qnn.QuantConv2d` & `qnn.QuantLinear` with all quantizers set to `None` even in the `quant_identity` layer and got validation accuracy of around 77% right?

**Comment by wilfredkisku** [comment]
What I meant was I used the `nn.Conv2d` and `nn.Linear` layers and got the above accuary.

**Comment by MohamedA95** [comment]
Try to take it step by step first build the model using qnn layers without any quantization, then add quantization step by step weight @ 8bit then input @ 8bit

**Comment by MohamedA95** [comment]
you can do that by setting `weight_quant=None` 
` self.conv1 = qnn.QuantConv2d(3, 16, 3, weight_quant=None, padding=1)`, The other quantizers are set to `None` by default.

**Comment by wilfredkisku** [comment]
Do I also need to make any changes to the training code? Thanks.

**Comment by wilfredkisku** [comment]
Thank you for being patient with the issue that I have been facing, I am a beginner and would like your help to be correct if there is a mistake that I am making. I am including the code, also the training part that might help you more in finding if I am making any mistakes. While training I can see that the weights are getting saved as it shows that there is a decrease in validation loss but while printing the loss value it shows the same values for consecutive epochs. I tried going step by step and switching off and on the quantization feature but it seems that as soon as I use layers related to the quantization library (brevitas) this issue pops up for me.

```
import torch.optim as optim

# specify loss function (categorical cross-entropy)
criterion = nn.CrossEntropyLoss()

# specify optimizer
optimizer = optim.Adam(net.parameters(), lr=0.001, betas=(0.9,0.999),eps=1e-08,weight_decay=0,amsgrad=False)
```

```
import torch.nn as nn
import torch.nn.functional as F
import brevitas.nn as qnn
from brevitas.quant import Int8Bias as BiasQuant
# define the CNN architecture
class QNet(nn.Module):
    def __init__(self):
        super(QNet, self).__init__()
        # convolutional layer (sees 32x32x3 image tensor)
        self.conv1 = qnn.QuantConv2d(3, 16, 3, padding=1, weight_quant=None,)
        # convolutional layer (sees 16x16x16 tensor)
        self.conv2 = qnn.QuantConv2d(16, 32, 3, padding=1, weight_quant=None,)
        # convolutional layer (sees 8x8x32 tensor)
        self.conv3 = qnn.QuantConv2d(32, 64, 3, padding=1, weight_quant=None,)
        # max pooling layer
        self.pool = nn.MaxPool2d(2, 2)
        # linear layer (64 * 4 * 4 -> 500)
        self.fc1 = qnn.QuantLinear(64 * 4 * 4, 500, bias=True)
        # linear layer (500 -> 10)
        self.fc2 = qnn.QuantLinear(500, 10, bias=True)
        # dropout layer (p=0.25)
        #self.dropout = nn.Dropout(0.25)

    def forward(self, x):
        # add sequence of convolutional and max pooling layers
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = self.pool(F.relu(self.conv3(x)))
        # flatten image input
        x = x.view(-1, 64 * 4 * 4)
        # add dropout layer
        #x = self.dropout(x)
        # add 1st hidden layer, with relu activation function
        x = F.relu(self.fc1(x))
        # add dropout layer
        #x = self.dropout(x)
        # add 2nd hidden layer, with relu activation function
        x = self.fc2(x)
        return x

# create a complete CNN
qnet = QNet()
print(qnet)

# move tensors to GPU if CUDA is available
if train_on_gpu:
    qnet.cuda()
```

```
# number of epochs to train the model
n_epochs = 30

valid_loss_min = np.Inf # track change in validation loss

for epoch in range(1, n_epochs+1):

    # keep track of training and validation loss
    train_loss = 0.0
    valid_loss = 0.0
    
    ###################
    # train the model #
    ###################
    qnet.train()
    for data, target in train_loader:
        # move tensors to GPU if CUDA is available
        if train_on_gpu:
            data, target = data.cuda(), target.cuda()
        # clear the gradients of all optimized variables
        optimizer.zero_grad()
        # forward pass: compute predicted outputs by passing inputs to the model
        output = qnet(data)
        # calculate the batch loss
        loss = criterion(output.to(torch.float32), target)
        # backward pass: compute gradient of the loss with respect to model parameters
        loss.backward()
        # perform a single optimization step (parameter update)
        optimizer.step()
        # update training loss
        train_loss += loss.item()*data.size(0)
        
    ######################    
    # validate the model #
    ######################
    qnet.eval()
    for data, target in valid_loader:
        # move tensors to GPU if CUDA is available
        if train_on_gpu:
            data, target = data.cuda(), target.cuda()
        # forward pass: compute predicted outputs by passing inputs to the model
        output = qnet(data)
        # calculate the batch loss
        loss = criterion(output.to(torch.float32), target)
        # update average validation loss 
        valid_loss += loss.item()*data.size(0)
    
    # calculate average losses
    train_loss = train_loss/len(train_loader.sampler)
    valid_loss = valid_loss/len(valid_loader.sampler)
        
    # print training/validation statistics 
    print('Epoch: {} \tTraining Loss: {:.6f} \tValidation Loss: {:.6f}'.format(
        epoch, train_loss, valid_loss))
    
    # save model if validation loss has decreased
    if valid_loss <= valid_loss_min:
        print('Validation loss decreased ({:.6f} --> {:.6f}).  Saving model ...'.format(
        valid_loss_min,
        valid_loss))
        torch.save(net.state_dict(), 'model_cifar_quant.pt')
        valid_loss_min = valid_loss
```
The training result for 20 epochs:
```
Epoch: 1 	Training Loss: 2.304736 	Validation Loss: 2.304074
Validation loss decreased (inf --> 2.304074).  Saving model ...
Epoch: 2 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 3 	Training Loss: 2.304736 	Validation Loss: 2.304074
Validation loss decreased (2.304074 --> 2.304074).  Saving model ...
Epoch: 4 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 5 	Training Loss: 2.304736 	Validation Loss: 2.304074
Validation loss decreased (2.304074 --> 2.304074).  Saving model ...
Epoch: 6 	Training Loss: 2.304736 	Validation Loss: 2.304074
Validation loss decreased (2.304074 --> 2.304074).  Saving model ...
Epoch: 7 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 8 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 9 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 10 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 11 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 12 	Training Loss: 2.304736 	Validation Loss: 2.304074
Validation loss decreased (2.304074 --> 2.304074).  Saving model ...
Epoch: 13 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 14 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 15 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 16 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 17 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 18 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 19 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 20 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 21 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 22 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 23 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 24 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 25 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 26 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 27 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 28 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 29 	Training Loss: 2.304736 	Validation Loss: 2.304074
Epoch: 30 	Training Loss: 2.304736 	Validation Loss: 2.304074
```

**Comment by wilfredkisku** [comment]
Thank you for your replies, I found that there was a bug in my code itself. Now, it works fine. Thanks again.
